solr/core/src/java/org/apache/solr/cli/StreamTool.java

```diff
@@ -0,0 +1,507 @@
+import org.apache.solr.client.solrj.io.stream.StreamContext;
+import org.apache.solr.client.solrj.io.stream.TupleStream;
+import org.apache.solr.client.solrj.io.stream.expr.Explanation;
+import org.apache.solr.client.solrj.io.stream.expr.Expressible;
+import org.apache.solr.client.solrj.io.stream.expr.StreamExpression;
+import org.apache.solr.client.solrj.io.stream.expr.StreamExpressionParser;
+import org.apache.solr.client.solrj.io.stream.expr.StreamFactory;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.handler.CatStream;
+
+/** Supports stream command in the bin/solr script. */
+public class StreamTool extends ToolBase {
+
+  public StreamTool() {
+    this(CLIO.getOutStream());
+  }
+
+  public StreamTool(PrintStream stdout) {
+    super(stdout);
+  }
+
+  private final SolrClientCache solrClientCache = new SolrClientCache();
+
+  @Override
+  public String getName() {
+    return "stream";
+  }
+
+  private static final Option EXECUTION_OPTION =
+          Option.builder("e")
+                  .longOpt("execution")
+                  .hasArg()
+                  .argName("CONTEXT")
+                  .required(false)
+                  .desc(
+                          "Execution context is either 'local' (i.e CLI process) or 'solr'. Default is 'solr'")
+                  .build();
+
+  private static final Option COLLECTION_OPTION = Option.builder("c")
+            .longOpt("name")
+            .argName("NAME")
+            .hasArg()
+            .desc(
+                "Name of the collection to execute on if workers are 'solr'.  Required for 'solr' worker.")
+            .build();
+
+  private static final Option FIELDS_OPTION = Option.builder("f")
+            .longOpt("fields")
+            .argName("FIELDS")
+            .hasArg()
+            .required(false)
+            .desc(
+                "The fields in the tuples to output. (defaults to fields in the first tuple of result set).")
+            .build();
+
+  private static final Option HEADER_OPTION =
+          Option.builder()
+                  .longOpt("header")
+                  .required(false)
+                  .desc("Whether or not to include a header line. (default=false)")
+                  .build();
+  private static final Option DELIMITER_OPTION = Option.builder()
+           .longOpt("delimiter")
+            .argName("CHARACTER")
+            .hasArg()
+            .required(false)
+            .desc("The output delimiter. (default=tab).")
+            .build();
+  private static final Option ARRAY_DELIMITER_OPTION = Option.builder()
+                .longOpt("array-delimiter")
+            .argName("CHARACTER")
+            .hasArg()
+            .required(false)
+            .desc("The delimiter multi-valued fields. (default=|)")
+            .build();
+
+  @Override
+  public Options getOptions() {
+    Options opts =
+            super.getOptions()
```

Comment:
I think you could simplify it by directly returning the value of `opts` here. Prevents any "surprises" further down this method.

Reply:
we could.. yet we have this pattern everywehre!